src/party.rs

```diff
@@ -24,67 +62,876 @@ pub struct BallotConfig {
+            }
+            if self.event_sender.send(PartyEvent::Finalize).is_err() {
+                self.status = PartyStatus::Failed;
+                return Err(BallotError::Communication(
+                    "Failed to send Finalize event".into(),
+                ));
+            }
+        }
+
+        Ok(self.get_value_selected())
+    }
+
+    /// Prepare state before running a ballot
+    fn prepare_next_ballot(&mut self) {
+        self.status = PartyStatus::None;
+        self.ballot += 1;
+
+        // Clean state
+        self.parties_voted_before = HashMap::new();
+        self.messages_1b_weight = 0;
+        self.value_2a = None;
+        self.messages_2av_senders = HashSet::new();
+        self.messages_2av_weight = 0;
+        self.messages_2b_senders = HashSet::new();
+        self.messages_2b_weight = 0;
+
+        // Cleaning channels
+        while self.event_receiver.try_recv().is_ok() {}
+        while self.msg_in_receiver.try_recv().is_ok() {}
+
+        self.status = PartyStatus::Launched;
     }
 
     /// Update party's state based on message type.
-    fn update_state(&mut self, msg: Vec<u8>, routing: MessageRouting) {
-        // TODO: Implement according to protocol rules.
+    fn update_state(&mut self, m: Vec<u8>, routing: MessageRouting) -> Result<(), BallotError> {
         match routing.msg_type {
-            ProtocolMessage::Msg1a => {}
-            ProtocolMessage::Msg1b => {}
-            ProtocolMessage::Msg1c => {}
-            ProtocolMessage::Msg2a => {}
-            ProtocolMessage::Msg2av => {}
-            ProtocolMessage::Msg2b => {}
+            ProtocolMessage::Msg1a => {
+                let msg: Message1aContent = serde_json::from_slice(m.as_slice())
+                    .map_err(|_| BallotError::MessageParsing("Failed to parse Msg1a".into()))?;
+
+                if msg.ballot != self.ballot {
+                    return Err(BallotError::InvalidState(
+                        "Ballot number mismatch in Msg1a".into(),
+                    ));
+                }
+
+                if routing.sender != self.cfg.leader {
+                    return Err(BallotError::InvalidState("Invalid leader in Msg1a".into()));
+                }
+
+                self.status = PartyStatus::Passed1a;
+            }
+            ProtocolMessage::Msg1b => {
+                let msg: Message1bContent = serde_json::from_slice(m.as_slice())
+                    .map_err(|_| BallotError::MessageParsing("Failed to parse Msg1b".into()))?;
+
+                if msg.ballot != self.ballot {
+                    return Err(BallotError::InvalidState(
+                        "Ballot number mismatch in Msg1b".into(),
+                    ));
+                }
+
+                if let Some(last_ballot_voted) = msg.last_ballot_voted {
+                    if last_ballot_voted >= self.ballot {
+                        return Err(BallotError::InvalidState(
+                            "Received outdated 1b message".into(),
```

Comment:
Is the message actually outdated in this case, or has the local instance fallen behind (potentially)?

Reply:
I believe there may be both these cases, however it doesn't impact on solution

Reply:
No it doesn't impact the solution, just want to make sure the error is correct for the user :)